lib/intern.rs

```diff
@@ -0,0 +1,115 @@
+extern crate lazy_static;
+
+pub use self::lazy_static::lazy_static;
+use std::marker;
+use std::thread;
+use std::vec;
+use std::collections;
+use std::sync;
+use serde;
+use differential_datalog::record::Record;
+
+type Symbol = u32;
+
+#[derive(Eq, PartialOrd, PartialEq, Ord, Clone, Hash)]
+pub struct intern_IObj<T> {
+    pub x: Symbol,
+    phantom: marker::PhantomData<T>
+}
+
+struct Interner<T> {
+    vec: vec::Vec<T>,
```

Comment:
why do you need the vector?
isn't a Map<T, T> enough?

Reply:
Sorry, I don't understand how `Map<T,T>` is enough.  We need `Map<T, Symbol>` to go from `T` to number and `Vec<T>` to go from number to `T`.